python/shape.cpp

```diff
@@ -40,12 +40,33 @@ template <class T> void bind_concatenate(py::module &m) {
 template <class T> void bind_reshape(pybind11::module &mod) {
   mod.def(
       "reshape",
-      [](const T &self, const std::vector<Dim> &labels,
-         const py::tuple &shape) {
-        Dimensions dims(labels, shape.cast<std::vector<scipp::index>>());
-        return reshape(self, dims);
+      [](const T &self, const py::dict &dims) {
+        Dimensions new_dims;
+        for (const auto &item : dims)
+          new_dims.addInner(item.first.cast<Dim>(),
+                            item.second.cast<scipp::index>());
+        return reshape(self, new_dims);
       },
-      py::arg("x"), py::arg("dims"), py::arg("shape"));
+      py::arg("x"), py::arg("dims"));
+}
+
+template <class T> void bind_reshape_data_array(pybind11::module &mod) {
+  mod.def(
+      "reshape",
+      [](const T &self, const Dim dim, const py::dict &to_dims) {
+        Dimensions new_dims;
+        for (const auto &item : to_dims)
+          new_dims.addInner(item.first.cast<Dim>(),
+                            item.second.cast<scipp::index>());
+        return reshape(self, dim, new_dims);
+      },
+      py::arg("x"), py::arg("dim"), py::arg("to_dims"));
+  mod.def(
+      "reshape",
+      [](const T &self, const py::list &from_dims, const Dim &to_dim) {
+        return reshape(self, from_dims.cast<std::vector<Dim>>(), to_dim);
```

Comment:
Why using `py::list` instead of `std::vector<std::string>`?

Reply:
no particular reason, i can change

Reply:
Note that if you cast yourself you cannot directly release the GIL (you have to use an explicit guard), whereas if pybind does the cast for you, you can.